File: bot/bot.py
import asyncio
import logging
from datetime import datetime
from aiogram import Bot, Dispatcher, F
from aiogram.filters import Command
from aiogram.types import Message, InlineKeyboardMarkup, InlineKeyboardButton, CallbackQuery
from supabase import create_client, Client

# Настрой свои данные от Supabase и Telegram Bot Token
TELEGRAM_TOKEN = "8845943428:AAEzxL-ITZyCd02fO9aMGlCMcE58uvM7vSw"
SUPABASE_URL = "https://dbodxldqvhtyjsdqppdh.supabase.co"
SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImRib2R4bGRxdmh0eWpzZHFwcGRoIiwicm9sZSI6ImFub24iLCJpYXQiOjE3OTAyMTg4MDEsImV4cCI6MjEwNTc5NDgwMX0.hXWWrb4Du2ec7j2yXnO7SrZWzwtmRCF4gBHpkK2_neY"

# Инициализация клиентов
bot = Bot(token=TELEGRAM_TOKEN)
dp = Dispatcher()
supabase: Client = create_client(SUPABASE_URL, SUPABASE_KEY)

logging.basicConfig(level=logging.INFO)


# Получение списка групп из Supabase для клавиатуры
async def get_groups_keyboard():
    response = supabase.table("groups").select("id, name").execute()
    keyboard = []
    for group in response.data:
        keyboard.append([
            InlineKeyboardButton(
                text=group["name"],
                callback_data=f"group_{group['id']}"
            )
        ])
    return InlineKeyboardMarkup(inline_keyboard=keyboard)


# Команда /start или выбор группы
@dp.message(Command("start"))
async def cmd_start(message: Message):
    keyboard = await get_groups_keyboard()
    await message.answer("Привет! Выбери свою группу для просмотра расписания:", reply_markup=keyboard)


# Обработка выбора группы через инлайн-кнопки
@dp.callback_query(F.data.startswith("group_"))
async def process_group_selection(callback: CallbackQuery):
    group_id = callback.data.split("_")[1]
    # Сохраняем выбор пользователя (можно в памяти или сессии, для примера выведем подтверждение)
    await callback.message.answer(
        f"Группа выбрана! Теперь используй команды:\n/today — расписание на сегодня\n/now — текущая пара")
    await callback.answer()


# Команда /today (расписание на текущий день недели)
@dp.message(Command("today"))
async def cmd_today(message: Message):
    # Определяем день недели (1 — Пн, ... 6 — Сб)
    # datetime.weekday(): 0 - Пн, 5 - Сб, 6 - Вс
    today_num = datetime.now().weekday() + 1

    if today_num > 6:
        await message.answer("Сегодня воскресенье, пар нет! 🎉")
        return

    # Запрос расписания из Supabase (для примера берем группу с id=1, ПО-42)
    response = supabase.table("schedule") \
        .select("*") \
        .eq("group_id", 1) \
        .eq("day_of_week", today_num) \
        .order("lesson_number") \
        .execute()

    schedule = response.data
    if not schedule:
        await message.answer("На сегодня пар в базе не найдено.")
        return

    text = f"📖 **Расписание на сегодня (День недели: {today_num}):**\n\n"
    for item in schedule:
        text += f"*{item['lesson_number']} пара* ({item['time_start']} - {item['time_end']})\n"
        text += f"📚 {item['subject_name']}\n\n"

    await message.answer(text, parse_mode="Markdown")


# Команда /now (текущая пара)
@dp.message(Command("now"))
async def cmd_now(message: Message):
    now_time = datetime.now().strftime("%H:%M:%S")
    today_num = datetime.now().weekday() + 1

    # Ищем пару, которая идет прямо сейчас
    response = supabase.table("schedule") \
        .select("*") \
        .eq("group_id", 1) \
        .eq("day_of_week", today_num) \
        .lte("time_start", now_time) \
        .gte("time_end", now_time) \
        .execute()

    current_lesson = response.data
    if not current_lesson:
        await message.answer("Сейчас никакая пара не идет ☕")
        return

    item = current_lesson[0]
    text = f"⚡ **Сейчас идет пара:**\n\n" \
           f"*{item['lesson_number']} пара* ({item['time_start']} - {item['time_end']})\n" \
           f"📚 {item['subject_name']}"

    await message.answer(text, parse_mode="Markdown")


async def main():
    await dp.start_polling(bot)


if __name__ == "__main__":
    asyncio.run(main())

File: bot/bot.js
const TelegramBot = require('node-telegram-bot-api');
const { createClient } = require('@supabase/supabase-js');

// ==============================
// Подключение к Supabase
// ==============================

const supabase = createClient(
  process.env.SUPABASE_URL,
  process.env.SUPABASE_ANON_KEY
);

// ==============================
// Запуск Telegram-бота
// ==============================

const token = process.env.TELEGRAM_BOT_TOKEN;

const bot = new TelegramBot(token, {
  polling: true
});

console.log('Telegram-бот успешно запущен!');

// ==============================
// Команды /start и /groups
// ==============================

bot.onText(/\/start|\/groups/, async (msg) => {
  const chatId = msg.chat.id;

  try {
    // Получаем список групп из Supabase
    const { data: groups, error } = await supabase
      .from('groups')
      .select('id, name')
      .order('id', { ascending: true });

    if (error) {
      console.error('Ошибка Supabase:', error);

      await bot.sendMessage(
        chatId,
        'Ошибка при получении групп: ' + error.message
      );

      return;
    }

    if (!groups || groups.length === 0) {
      await bot.sendMessage(
        chatId,
        'В базе данных пока нет групп.'
      );

      return;
    }

    // Создаём кнопки групп
    const keyboard = groups.map(group => [
      {
        text: group.name,
        callback_data: `group_${group.id}`
      }
    ]);

    await bot.sendMessage(
      chatId,
      '📚 Выберите вашу группу:',
      {
        reply_markup: {
          inline_keyboard: keyboard
        }
      }
    );

  } catch (err) {
    console.error('Ошибка /start:', err);

    await bot.sendMessage(
      chatId,
      'Произошла непредвиденная ошибка.'
    );
  }
});

// ==============================
// Нажатие на кнопку группы
// ==============================

bot.on('callback_query', async (query) => {
  const chatId = query.message.chat.id;
  const data = query.data;

  // Убираем "часики" на кнопке после нажатия
  try {
    await bot.answerCallbackQuery(query.id);
  } catch (err) {
    console.error('Ошибка callback:', err);
  }

  if (!data || !data.startsWith('group_')) {
    return;
  }

  const groupId = data.replace('group_', '');

  try {
    // ==============================
    // Получаем расписание группы
    // ==============================

    const { data: schedule, error } = await supabase
      .from('schedule')
      .select('*')
      .eq('group_id', groupId)
      .order('day_of_week', { ascending: true })
      .order('lesson_number', { ascending: true });

    if (error) {
      console.error('Ошибка Supabase:', error);

      await bot.sendMessage(
        chatId,
        'Ошибка при получении расписания: ' + error.message
      );

      return;
    }

    if (!schedule || schedule.length === 0) {
      await bot.sendMessage(
        chatId,
        'Для этой группы пока нет расписания.'
      );

      return;
    }

    // ==============================
    // Разделяем расписание по дням
    // ==============================

    const days = {};

    schedule.forEach(row => {
      if (!days[row.day_of_week]) {
        days[row.day_of_week] = [];
      }

      days[row.day_of_week].push(row);
    });

    // ==============================
    // Названия дней недели
    // ==============================

    const dayNames = {
      1: 'Понедельник',
      2: 'Вторник',
      3: 'Среда',
      4: 'Четверг',
      5: 'Пятница',
      6: 'Суббота',
      7: 'Воскресенье'
    };

    // ==============================
    // Отправляем каждый день отдельно
    // ==============================

    for (const [day, lessons] of Object.entries(days)) {

      const dayName = dayNames[day] || `День ${day}`;

      let messageText = `📅 ${dayName}\n\n`;

      lessons.forEach(row => {

        messageText += `🔹 Пара №${row.lesson_number}\n`;

        messageText +=
          `📚 Предмет: ${row.subject_name || 'Не указан'}\n`;

        if (row.time_start && row.time_end) {
          // Убираем секунды: 08:30:00 → 08:30
          const startTime = row.time_start.slice(0, 5);
          const endTime = row.time_end.slice(0, 5);

          messageText +=
            `🕐 Время: ${startTime} — ${endTime}\n`;
        }

        messageText += '\n';
      });

      await bot.sendMessage(chatId, messageText);
    }

  } catch (err) {
    console.error('Ошибка при загрузке расписания:', err);

    await bot.sendMessage(
      chatId,
      'Ошибка при загрузке расписания.'
    );
  }
});

// ==============================
// Обработка ошибок polling
// ==============================

bot.on('polling_error', (error) => {
  console.error('Polling error:', error.message);
});